Rfm 17.1 - Sharing Provider Records with Multiaddress

cortze opened the pull request:
This is the first draft of the report that extends [RFM17](https://github.com/protocol/network-measurements/blob/master/results/rfm17-provider-record-liveness.md) to measure if the Multiaddresses of a content provider are being shared during the retrieval process of a CID process. 

It includes the study's motivation, the methodology we followed, the discussion of the results we got out of the study, and a conclusion.

All kind of feedback is appreciated, so please, go ahead to point out improvements! 

Also, should I be running a more extensive set of CIDs for extended periods? 

cc: @yiannisbot @guillaumemichel @dennis-tra

cortze commented on RFMs.md:
```diff
@@ -615,6 +617,33 @@ Additional context can be found [here](https://pl-strflt.notion.site/Provider-Re
+
+However, If this process of walking the DHT looking for the PR succeeds, the `kubo` client will get the link between the CID and the PeerID that host the content. Thus, the user still has to make a second DHT lookup to find the latest public multiaddress of that specific peer. 
+
+Each public multiaddress for any peer in the network has assigned a Time To Live (TTL) duration, which can vary between `go-ipfs` or  `kubo` versions. It was initially set to 10 mins but was incremented to 30 mins in the `go-libp2p@v0.22.0` update on August 18, 2022. In some occasions, if the user fetches the PRs inside the time window where the multiaddress of the provider didn't expire, the pro ...
```
The changes were proposed in this [PR](https://github.com/libp2p/go-libp2p/commit/c2821791bac7d638890accc98798bf4cbfe122e7) in 2021 by Stebalien. But it wasn't until August 18 this year that it wasn't [merged](https://github.com/libp2p/go-libp2p/pull/1683) to the `go-libp2p` master branch.
They are hardcoded parameters in the `go-libp2p` code base, so previous versions to that merge are set to 10mins TTL

cortze commented on RFMs.md:
```diff
@@ -615,6 +617,33 @@ Additional context can be found [here](https://pl-strflt.notion.site/Provider-Re
+
+Each public multiaddress for any peer in the network has assigned a Time To Live (TTL) duration, which can vary between `go-ipfs` or  `kubo` versions. It was initially set to 10 mins but was incremented to 30 mins in the `go-libp2p@v0.22.0` update on August 18, 2022. In some occasions, if the user fetches the PRs inside the time window where the multiaddress of the provider didn't expire, the pro ...
+
+This RFM, which is an extension of the RFM17 for its close relation to the PR retrievability aspect, aims to measure whether the shared PRs for a given CID actually contain the multiaddress of the provider and for how long they are shared. The final intention of the RFM is to discuss whether we can avoid this second DHT lookup by increasing the TTL of the multiaddress linked to the PR to 24h (sam ...
```
Nice idea yes, thanks Yiannis!

cortze commented on RFMs.md:
```diff
@@ -615,6 +617,33 @@ Additional context can be found [here](https://pl-strflt.notion.site/Provider-Re
+#### Measurement Plan
+
+- Spin up a node that generates random CIDs and publishes provider records.
+- Periodically attempt to fetch the PR from the DHT, tracking whether they are retrievable and whether they are shared among the multiaddresses.
```
> What do we mean "shared among the multiaddresses"? Whether the PR can be found in the multiaddress of the original node that stored the record?

In the networking layer, when we ask for the PR of a CID, we just get as a reply an `AddrInfo` of each provider that the remote peer is aware of. So the PR as we understand it, it's just how we store it in the DHT. 
This `AddrInfo` contains two fields: `PeerID` and `Multiaddresses`, and it will only fill up the `Multiaddresses` if their TTL are still valid.

>  how about just getting a bunch of PeerIDs and their multiaddresses and pinging them over time to see whether they listen to that multiaddress?

That can be a nice side experiment, yes. Although I think that we are indirectly doing it. In the hoarder, I keep the `AddrInfo` of each PR Holder with the first `Multiaddresses` that I got from the publication, and I only use those addresses to establish the connections. So if they were changing IPs, I wouldn't be able to connect to them. 

Let me know anyways if you want me to set up a specific test for the IP rotation.

cortze replied:
> we either mean if the PR is available in the multiaddresses of all the content providers, or if all the multiaddresses of all content providers are included in the PR. :) Is it any of these two?

We are in the second one, we only get an `AddrInfo` for those providers that the remote peer is aware of, and it depends on the TTL of the `Multiaddress` to include them or not in the `AddrInfo` of the provider. Should I say it in the opposite way "the multiaddress is shared among the PRs"?
Let me point you to the code; maybe is easier to understand it:
- [here](https://github.com/libp2p/go-libp2p-kad-dht/blob/e33a4be6e9a3a8fb603d21126e2d8a42c5e37d1b/routing.go#L536) is the inner method during the `dht.FindProviders()` method.
 - [here](https://github.com/libp2p/go-libp2p-kad-dht/blob/e33a4be6e9a3a8fb603d21126e2d8a42c5e37d1b/pb/protocol_messenger.go#L137) is the networking method to ask for the PRs to a remote peer.

> Can we run an experiment where we keep those connections open for a time period equal to the Expiry Interval?
Absolutely! I can make a new run with 10k-20k CIDs over 60h if that is enough

cortze commented on results/rfm17.1-sharing-prs-with-multiaddresses.md:
```diff
@@ -0,0 +1,124 @@
+
+Results are similar when we analyze the replies of the peers that report back the PR from the DHT lookup process. We increased the number of content providers we were looking for to track the multiple remote peers. Figure [3] represents the number of remote peers reporting the PR for the CIDs we were looking for, where we can see a stable 20 peers by median over the entire study. 
+
+For those wondering why more than 20 peers (k replication value when publishing the CIDs) are reporting the PR, we must remind you that Hydra-Boosters share the PR database among different `PeerID` heads. Which means that if one hydra hears about a PR, all the leaders of that common database will also share it.  
```
In the hoarder, I always check if the PRs that are shared back match the `PeerID` of the content publisher (which, in this case, is my local host 1, the publisher one). So if someone tries to reprovide or ping the CID, it wouldn't affect these results. 

About the hydras, I'm not aware of how many hydra "bellies" are out there. Is there a single big one or multiple small ones? Also, we have to keep in mind that the DHT lookup converges into a region of the SHA256 hash space, so it's quite unlucky that we will get connections and replies from hydras that are in the opposite part of the hash space.

cortze commented on results/rfm17.1-sharing-prs-with-multiaddresses.md:
```diff
@@ -0,0 +1,124 @@
+
+It is essential to notice that despite being able to show empirically what we already knew, the combo of `PeerID` + `Multiaddress` gets shared only over the TTL of the `Multiaddress` records, the result that we got from the DHT lookup process doesn't fully match the previous results. 
+
+We would expect to retrieve the combo over the current TTL of the multiaddress records. However, the fragmentation between IPFS nodes in the network makes some of the nodes only reply with the `PeerID` of the content's provider, and in the current [`libp2p/go-libp2p-kad-dht`](https://github.com/libp2p/go-libp2p-kad-dht) implementation, the reply of the lookup only gets the `AddrInfo` of the first ...
```
I wonder how Grammarly didn't catch this... :))
Rephasing it in the next iteration, though!

cortze commented on results/rfm17.1-sharing-prs-with-multiaddresses.md:
```diff
@@ -0,0 +1,124 @@
+
+With this study, we have demonstrated empirically that the combo of `PeerID` + `Multiaddress` are, in fact, shared as long as the TTL of the `Multiaddress` records don't expire. This means that if we increase this TTL to match the PR expiration time, we could be reducing to a single DHT lookup the process of retrieving a CID's content using the public DHT.
+
+On the other hand, we have identified some code limitations that could affect the impact of such improvements as far as the network keeps fragmented based on client versions and mismatches between configurations.
```
> Is that because of DHT Server nodes not having upgraded their version?

Mostly yes, the major limiting factor here is the wide variety of versions. Which limits the expected result (PeerID + Multiaddress for the whole TTL ) only for the minimum TTL in the network (10 mins in this case).

cortze commented:
Thanks for the feedback @yiannisbot , I really appreciate it! 

> My main worry is that the claim of: "if a multiaddress is returned together with the PeerID for the TTL period (10mins or 30mins), then we can extend the TTL to the PR expiry interval" doesn't really hold. Why would we arrive to this conclusion

I will try to make it a bit more explicit in the conclusion (my bad). It's not an "it won't hold" statement. It is an "It won't have as much impact as we are expecting" statement.

As far as your network has different TTL values for Multiaddresses (like in the current network), the smallest TTL will be the one limiting negatively the final result of the DHT lookup process (at least the `go-libp2p-kad-dht` one). So unless the largest part of the network updates to that TTL, we will still face the same problem, and there will still be sporadic problems originated from those remaining "old" clients. (The double-hashing implementation would be a nice incentive to force a total network update)

> Basically, similar to the CID Hoarder, what we need here is a PeerID Hoarder :-D This tool would get a lot of PeerIDs, record the multiaddress by which we first saw the peer and then periodically ping the peer to figure out if it changed its Multiaddress within the PR Expiry Interval.

I left you a [comment](https://github.com/protocol/network-measurements/pull/22#discussion_r1022570651) as well in the [#22 comment](https://github.com/protocol/network-measurements/pull/22#discussion_r1022489894)
I think that we have a few options here. The hoarder already does this indirectly (it contacts the PR Holders to the `Multiaddress` that we stored while storing the PRs). Also, I think that Nebula already tracks IP rotation. We could have a deeper chat about this :)

I'll iter again over your comments and suggestions, will ping you back whenever I make a commit!

dennis-tra commented:
> I'm not sure if this functionality can easily be included in Nebula @dennis-tra ?

Sorry for the late reply! The information is already recorded by Nebula and would just need to be analyzed :)

> I'll iter again over your comments and suggestions, will ping you back whenever I make a commit!

Just ping here or in Discord and I'll also have a proper read. I just skimmed it in the past 🙈

cortze commented:
I already added some explanations and most of the changes that @yiannisbot suggested. I set up another Hoarder run with 20k CIDs for 60 hours, so the plots and some numbers might change.

If you can go through and give me some thoughts @dennis-tra , I would appreciate your feedback as well 😄

guillaumemichel commented on results/rfm17.1-sharing-prs-with-multiaddresses.md:
```diff
@@ -0,0 +1,127 @@
+
+_Figure 2: Number of PR Holders replying with the `PeerID` + `Multiaddress` combo._
+
+### 4.2-Reply of peers reporting the PR during the DHT lookup
```
I am not sure I understood this correctly: The only difference between 4.1 and 4.2 is that Hydras appears in 4.2 but not 4.1?

cortze replied:
Since hydras are present in the set of PR holders, they appear in both 4.1 and 4.2.
However, since the DHT lookup wasn't stopped after the first retrieval of the PRs, I assume that most of the peers that report the PRs beyond those initial PR Holders are Hydras (for their shared DB of PR).

guillaumemichel replied:
So what exactly is the perform operation? Is it only a `FindProviders`? And it may get more than 20 peers responding with the PR, because some peers on the path to the CID would be Hydra nodes?
As the number of hops in a DHT lookup is usually 3-5, we would expect at MOST 23-25 peers responding with a PR, if all of the peers helping to route the request (NOT PR holders) are Hydra nodes. According to the plot in 4.2 there are regularly much more than this number. How do you explain this?
Or maybe I missed something here ^^

cortze replied:
> So what exactly is the perform operation? Is it only a FindProviders?

Yes, it's a modification of the `FindProviders()` method that doesn't look in the local Provider DB of the host, and that directly performs the DHT lookup.

> And it may get more than 20 peers responding with the PR, because some peers on the path to the CID would be Hydra nodes?

Exactly, that is the explanation that I gave for this phenomenon.

> As the number of hops in a DHT lookup is usually 3-5, we would expect at MOST 23-25 peers responding with a PR

Can you give a bit more context on this statement? My understanding from RFM 17 is that we perform [between 3 and 6 hops](https://github.com/protocol/network-measurements/blob/master/implementations/rfm17-provider-record-liveness/plots/k20/total_hops.png), however, that only determines the depth of the peer tree that is built during the lookup. We are not taking into account that the tree can also grow in width.

cortze replied:
@yiannisbot The `FindProviders()` that I use in the hoarder slightly differs from the vanilla operation: 
It removes the "[Find in the ProvidersStore](https://github.com/cortze/go-libp2p-kad-dht/blob/6b490320a6c1b70eba2031260a2515c26e7519fe/routing.go#L529)" operation, forcing it to look for the Providers only using the vanilla DHT lookup, and adds some traces to track when we receive a new Provider.

I've been relaunching the test with a two-minute timeout for the `FindProviders` operation, and the results seem to be in the range that @guillaumemichel suggests (keep in mind that the Hydras' DB has been plugged off). 

The number of remote peers replying with the PR during the DHT lookup (with a 2-minute timeout) looks like this.
![image](https://user-images.githubusercontent.com/45786396/213124862-a00a3613-b6a1-4853-9e03-6a1a727a2d95.png)

cortze commented:
> @cortze do we have any results from this experiment? I think with these results and addressing Guillaume's question, this should be ready to be merged, right?

@yiannisbot  The results of this run were not as good as I expected. To track such a large set of CIDs, I had to increase the concurrency parameters of the hoarder, and as we spotted in our last meeting (link to the [Issue](https://github.com/libp2p/go-libp2p-kad-dht/issues/805) describing the bottleneck) the code is not that prepared to support such a high degree of concurrency.

However, I think that even with such a low number of CIDs and a lower ping-interval between pings (3 minutes), we can conclude that [increasing provider Multiaddress' TTL](https://github.com/libp2p/go-libp2p-kad-dht/pull/795) would improve content fetching times. And the impact would be much higher if we merge it with [go-libp2p-kad-dht#802](https://github.com/libp2p/go-libp2p-kad-dht/pull/802).

RFM17 already proved that the IP rotation of PRHolders barely happens:
![image](https://user-images.githubusercontent.com/45786396/213138998-e5142bf4-69e7-4e20-98c4-05856eaf1c1b.png)

cortze commented:
@yiannisbot I've updated the document with your suggestions and with two extra paragraphs describing:
1. The observed IP-churn of DHT servers in IPFS (from the RFM-17)
2. Contribution section, where I aggregated all the pull requests related to this RFM 

I've also updated the figures. The new ones have the DHT lookup limited to 2 mins - which shows a reasonable number of peers that return the PRs as pointed out by @guillaumemichel .

The new data still faces a lower number of online PR Holders due to a problem storing the records in a part of the network. However, I consider them more than good enough to conclude that increasing the TTL of the Provider's Multiaddres would avoid the second DHT lookup to map the `PeerID` of the Provider with its `Multiaddres`.

Let me know what do you think about the update :) Cheers!
